### PR TITLE
Feature/cast

### PR DESCRIPTION
**Pull request by maniland**
adds:
- roku docs
- Chromecast web docs

### CODE REVIEW

File: docs/source/integrations/casting/chromecast.md [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+
+Typically the `Sender` would reside either as part of the web page or your chosen player.
+
+The `Receiver` is an application that you host and register with Google.
+
+Below is an example of basic `Sender` & `Receiver` side code using the Google SDK.  
+
+## Sender
+
+setup of the load function within your `sender` within your site or player.
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]:
```suggestion
Setup of the load function within your `sender` within your site or player.
```

File: docs/source/integrations/casting/chromecast.md [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+    request.currentTime = 0;
+    this._castContext.getCurrentSession().loadMedia(request).catch((error) => {
+        console.error(error);
+    });
+}
+```
+
+## Receiver
+
+setup of the host within your `receiver` application.
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]:
```suggestion
Setup of the host within your `receiver` application.
```

File: docs/source/integrations/casting/chromecast.md [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+```javascript
+if (event.data.media && event.data.media.contentId){
+    let host, protocol;
+    let url = event.data.media.contentId;
+    let initStart = event.data.media.currentTime || 0;
+    let autoplay = event.data.autoplay !== false;
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]:
```suggestion
let autoplay = !!event.data.autoplay;
```

File: docs/source/integrations/casting/chromecast.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+    let customData = event.data.media.customData || {};
+
+    host = new cast.player.api.Host({mediaElement:this._video, url});
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: Where does this come from?

File: docs/source/integrations/casting/chromecast.md [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    host.onError = (errorCode) => {
+        console.error('Fatal Error ' + errorCode);
+        this._unloadPlayer();
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: Same this here...?

File: docs/source/integrations/casting/roku.md [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+## Basic setup
+
+Follow the getting started instructions laid out in the [Roku documentation](https://developer.roku.com/en-gb/overview) to set up your environment. Roku uses it's own scripting language called BrightScript to set up channels/apps.
+
+We recommend using DASH with PlayReady on Roku devices.
+
+the documentation for which can be found in the [content protection](https://developer.roku.com/en-gb/docs/specs/content-protection.md)
+
+```javascript
+customData = "<VUDRM-TOKEN>";
```
Comment (stefanovualto) [inline docs/source/integrations/casting/roku.md]: What do you think about puttin the customData `<VUDRM-TOKEN>` value directly where it is use?

File: docs/source/integrations/casting/roku.md [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+
+ The Roku OS was purpose-built for streaming and runs across all Roku devices, including streaming players and Roku TVs. On the Roku platform the applications that stream your media are called channels.
+
+## Basic setup
+
+Follow the getting started instructions laid out in the [Roku documentation](https://developer.roku.com/en-gb/overview) to set up your environment. Roku uses it's own scripting language called BrightScript to set up channels/apps.
+
+We recommend using DASH with PlayReady on Roku devices.
+
+the documentation for which can be found in the [content protection](https://developer.roku.com/en-gb/docs/specs/content-protection.md)
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
The documentation for which can be found in the [content protection](https://developer.roku.com/en-gb/docs/specs/content-protection.md).
```

File: docs/source/integrations/casting/roku.md [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+## Testing your stream
+
+You can also test your streams outside of the Roku app using the [Stream testing tool](https://developer.roku.com/en-gb/docs/developer-program/dev-tools/tools-overview.md#stream-testing-tool)
+
+- Add your device `IP` and `Model` to the device manager.
+- Ensure that `Mode` is set to `Video`
+- Ensure that `Format` is set to either `Auto` or `DASH`
+- Within the DRM section choose `encoding type` : `PlayReady: LA And Challenge`
+- Put your newly generated VUDRM-token in to the custom data input box.
+- in `License server Url(optional)` put `https://playready-license.drm.technology/rightsmanager.asmx`
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
- In `License server Url(optional)` put `https://playready-license.drm.technology/rightsmanager.asmx`.
```

File: docs/source/integrations/casting/roku.md [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+- click `Play`
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
- Click `Play`.
```

File: docs/source/integrations/casting/chromecast.md [hunk 10]
```diff
@@ -0,0 +1,94 @@
+# Chromecast
+
+Chromecast is a casting device developed by Google, to enable streaming from web apps upon large screen display devices.
+
+Google have extensive [documentation](https://developers.google.com/cast/docs/developers) about their SDK'. If you have never developed a Chromecast application before we would strongly advise you start here.
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/chromecast.md]:
```suggestion
Google have extensive [documentation](https://developers.google.com/cast/docs/developers) about their SDK. If you have never developed a Chromecast application before we would strongly advise you start here.
```

File: docs/source/integrations/casting/roku.md [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+contentNode.encodingType = "PlayReadyLicenseAcquisitionAndChallenge"
+contentNode.encodingKey = "PlayReadyLicenseServerUrl" + "%%%" + customData
+```
+
+## Testing your stream
+
+You can also test your streams outside of the Roku app using the [Stream testing tool](https://developer.roku.com/en-gb/docs/developer-program/dev-tools/tools-overview.md#stream-testing-tool)
+
+- Add your device `IP` and `Model` to the device manager.
+- Ensure that `Mode` is set to `Video`
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
- Ensure that `Mode` is set to `Video`.
```

File: docs/source/integrations/casting/roku.md [hunk 12]
```diff
@@ -0,0 +1,32 @@
...
+- Ensure that `Format` is set to either `Auto` or `DASH`
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
- Ensure that `Format` is set to either `Auto` or `DASH`.
```

File: docs/source/integrations/casting/roku.md [hunk 13]
```diff
@@ -0,0 +1,32 @@
...
+- Within the DRM section choose `encoding type` : `PlayReady: LA And Challenge`
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
- Within the DRM section choose `encoding type` : `PlayReady: LA And Challenge`.
```

File: docs/source/integrations/casting/chromecast.md [hunk 14]
```diff
@@ -0,0 +1,99 @@
...
+
+Setup of the host within your `receiver` application.
+
+```javascript
+let videoElement = document.getElementById("your-video-element");
+if (event.data.media && event.data.media.contentId){
+    let host, protocol;
+    let url = event.data.media.contentId;
+    let initStart = event.data.media.currentTime || 0;
+let autoplay = !!event.data.autoplay;
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: indentation plz

File: docs/source/integrations/casting/chromecast.md [hunk 15]
```diff
@@ -0,0 +1,99 @@
...
+```
+
+## Receiver
+
+Setup of the host within your `receiver` application.
+
+```javascript
+let videoElement = document.getElementById("your-video-element");
+if (event.data.media && event.data.media.contentId){
+    let host, protocol;
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: put the declaration of host directly where do you asign it.

Comment (stefanovualto) [reply docs/source/integrations/casting/chromecast.md]: What is in protocol?

File: docs/source/integrations/casting/chromecast.md [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        }
+    }
+```
+
+### MPEG-DASH overrides
+
+```javascript
+let kid;
+host.protectionSystem = cast.player.api.ContentProtection.WIDEVINE;
+host.licenseUrl = customData.laUrl || 'https://widevine-proxy.drm.technology/proxy';
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]:
```suggestion
host.licenseUrl = 'https://widevine-proxy.drm.technology/proxy';
```

File: docs/source/integrations/casting/chromecast.md [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+host.processManifest = (manifest) => {
+    let parser = new DOMParser();
+    let xmlDoc = parser.parseFromString(manifest, 'text/xml');
+    let el = xmlDoc.querySelector('ContentProtection');
+    kid = el ? el.getAttribute('cenc:default_KID') : '';
+    return manifest;
+};
+
+host.updateLicenseRequestInfo = (requestInfo) => {
+    let token = customData.token;
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: where customData come from?

File: docs/source/integrations/casting/roku.md [hunk 18]
```diff
@@ -0,0 +1,33 @@
...
+contentNode = createObject("roSGNode", "contentNode")
+contentNode.streamFormat = "dash"
+contentNode.url = "<MPEG-DASH-URL>"
+contentNode.encodingType = "PlayReadyLicenseAcquisitionAndChallenge"
+contentNode.encodingKey = "PlayReadyLicenseServerUrl" + "%%%" + customData
+```
+
+## Testing your stream
+
+You can also test your streams outside of the Roku app using the [Stream testing tool](https://developer.roku.com/en-gb/docs/developer-program/dev-tools/tools-overview.md#stream-testing-tool)
```
Comment (PerritonBranch) [inline docs/source/integrations/casting/roku.md]:
```suggestion
You can also test your streams outside of the Roku app using the [Stream testing tool](https://developer.roku.com/en-gb/docs/developer-program/dev-tools/tools-overview.md#stream-testing-tool).
```

File: docs/source/integrations/casting/chromecast.md [hunk 19]
```diff
@@ -0,0 +1,100 @@
...
+Below is an example of basic `Sender` & `Receiver` side code using the Google SDK.  
+
+## Sender
+
+Setup of the load function within your `sender` within your site or player.
+
+```javascript
+load(url, token, laUrl) {
+    if (!this.isConnected) { return };
+    let mediaInfo, request;
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: Wouldn't be better to declare the variable next to their assignations?

File: docs/source/integrations/casting/chromecast.md [hunk 20]
```diff
@@ -0,0 +1,97 @@
...
+    let mediaInfo = new chrome.cast.media.MediaInfo(url);
+    mediaInfo.metadata = new chrome.cast.media.GenericMediaMetadata();
+    mediaInfo.metadata.metadataType = chrome.cast.media.MetadataType.GENERIC;
+    mediaInfo.metadata.title = 'VUDRM Demo';
+    mediaInfo.customData = { laUrl, token };
+
+    let request = new chrome.cast.media.LoadRequest(mediaInfo);
+    request.autoplay = true;
+    request.currentTime = 0;
+    this._castContext.getCurrentSession().loadMedia(request).catch((error) => {
```
Comment (stefanovualto) [inline docs/source/integrations/casting/chromecast.md]: Sorry what is `this._castContext`?

**Pull request merged**